File: mergesort.c
#include "mergesort.h"

void merge(int *arr, int start, int mid, int end)
{
    int n1 = mid - start + 1, n2 = end - mid, i, j;
    int *left = (int *)malloc(sizeof(int) * n1);
    int *right = (int *)malloc(sizeof(int) * n2);

    for (i = 0; i < n1; i++)
        left[i] = arr[i + start];

    for (i = 0; i < n2; i++)
        right[i] = arr[i + mid + 1];

    int k = start;
    i = j = 0;

    while (i < n1 && j < n2)
    {
        if (left[i] <= right[j])
            arr[k++] = left[i++];
        else
            arr[k++] = right[j++];
    }

    while (i < n1)
        arr[k++] = left[i++];

    while (j < n2)
        arr[k++] = right[j++];
}

void mergesort(int *arr, int start, int end)
{
    int mid = start + (end - start) / 2;
    if (start < end)
    {
        mergesort(arr, start, mid);
        mergesort(arr, mid + 1, end);
        merge(arr, start, mid, end);
    }
}

void mergesort_th(int *arr, int start, int end, int num_th)
{
    int t, rc;
    void *status;

    printf("1\n");

    struct m_thread_data td_arr[num_th];
    for (t = 0; t < num_th; t++)
    {
        td_arr[t].arr = arr;
        td_arr[t].start = t * ((end - start) / 4);
        td_arr[t].end = (t + 1) * ((end - start) / 4) - 1;
    }

    printf("2\n");

    pthread_t threads[num_th];
    for (t = 0; t < num_th; t++)
    {
        printf("Thread start: %d", t);
        rc = pthread_create(&threads[t], NULL, mergesort_th_worker,
                            (void *)&td_arr[t]);
        if (rc)
            printf("ERROR; return code from pthread_create() is %d\n", rc);
    }

    printf("3\n");

    for (t = 0; t < num_th; t++)
    {
        rc = pthread_join(threads[t], &status);
        if (rc)
            printf("ERROR; return code from pthread_join() is %d\n", rc);
    }
}

void *mergesort_th_worker(void *threadargs)
{
    struct m_thread_data *targs;
    targs = (struct m_thread_data *)threadargs;

    mergesort(targs->arr, targs->start, targs->end);

    pthread_exit(NULL);
}

File: python/mergesort.py
from threading import Thread
import numpy as np


def merge(arr, low, mid, high):
    left = np.zeros([mid - low + 1], dtype=int)
    right = np.zeros([high - mid], dtype=int)

    n1 = mid - low + 1
    n2 = high - mid

    for i in range(n1):
        left[i] = arr[i + low]

    for i in range(n2):
        right[i] = arr[i + mid + 1]

    k = low
    i = j = 0

    while i < n1 and j < n2:
        if left[i] <= right[j]:
            arr[k] = left[i]
            k += 1
            i += 1
        else:
            arr[k] = right[j]
            k += 1
            j += 1

    while i < n1:
        arr[k] = left[i]
        k += 1
        i += 1

    while j < n2:
        arr[k] = right[j]
        k += 1
        j += 1


def merge_sort(arr, low, high):
    mid = int(low + (high - low) // 2)
    if low < high:
        merge_sort(arr, low, mid)
        merge_sort(arr, mid + 1, high)
        merge(arr, low, mid, high)


def merge_sort_th_st(arr, low, high, num_th):
    size = high - low
    mid = int(low + size // 2)

    threads = []
    for i in range(num_th):
        low2 = i * int(size / num_th)
        high2 = (i + 1) * int(size / num_th) - 1

        th = Thread(target=lambda: merge_sort(arr, low2, high2))
        th.start()

        threads.append(th)

    for th in threads:
        th.join()

    if low < high:
        merge_sort(arr, low, mid)
        merge_sort(arr, mid + 1, high)
        merge(arr, low, mid, high)


def merge_sort_th_dy(arr, low, high):
    size = high - low
    mid = int(low + size // 2)

    if low < high:
        lthread = Thread(target=lambda: merge_sort(arr, low, mid))
        lthread.start()
        rthread = Thread(target=lambda: merge_sort(arr, mid + 1, high))
        rthread.start()
    
    lthread.join()
    rthread.join()
    
    merge(arr, low, mid, high)